src/providers/ogr/qgsogrconnpool.h

```diff
@@ -61,7 +61,13 @@ class QgsOgrConnPoolGroup : public QObject, public QgsConnectionPoolGroup<QgsOgr
     Q_OBJECT
 
   public:
-    QgsOgrConnPoolGroup( QString name ) : QgsConnectionPoolGroup<QgsOgrConn*>( name ) { initTimer( this ); }
+    QgsOgrConnPoolGroup( QString name ) : QgsConnectionPoolGroup<QgsOgrConn*>( name ), mRefCount( 0 ) { initTimer( this ); }
+    void ref() { ++mRefCount; }
```

Comment:
> AFAICS the provider has a separate connection (r/w and not acquired from the connection pool) and only the feature iterators acquire connections from the connection pool.

You are right that the provider does not actually use any of the connection pool connections, however, the `ref` and `unref` at provider level is a performance enhancement: if everytime a feature source is deleted, the pool is cleared, then we don't get the benefit anymore of keeping connections alive. On the other hand, if we keep the connections alive as long as the provider exists, then there is a good chance that we are re-using connections as often as possible.

> Each connection should only be use by one feature iterator at a time. So for parallel feature iterators there need to be separate connections - preferable unexpired connections that destroyed feature iterators returned to the pool or otherwise new ones.

This is true. But as pointed out above, the hope is to keep connections alive beyond the scope of a feature source.